### DISCUSSION TITLE
Diagram of control flow from a plugin's point of view

### DISCUSSION THREAD

**Discussion by oprypin** [mkdocs/mkdocs#2386] [Show and tell]
![mkdocs](https://user-images.githubusercontent.com/371383/116735918-79c28080-a9ef-11eb-8fd3-3ed80b1ecde0.png)
Spent a whole day on this :weary:

**Comment by oprypin** [comment]
Opened a pull request to perhaps add it to the docs: https://github.com/mkdocs/mkdocs/pull/2387

**Comment by Andre601** [comment]
Should the bottom part be separate?
Like is it really supposed to not be connected in some way?

Also, what actions lead to `on_pre_template -> template_name` and `build_template`? There is only an arrow with no origin.

**Reply by oprypin** [reply]
The items on the diagram have no promise of being connected, all others just happen to be connected because they have direct data flow between them. I could draw arrows from `on_config` to every single `config` entry, but it's not helpful. (the real connectivity, by the way, is ensured by the always-present `self`).
Instead, the vertical direction represents chronological sequence.
So according to my conventions, it is intended. If you have an alternative, let me know.

There is also no promise of "coming from somewhere". It is quite complicated how the theme templates get chosen, so I just omitted it. I could just connect it to `on_config` or something, but it makes the graph tool produce a weirder overall result.

That said, I can certainly see that everyone would be confused by this in the same way. Just need some constructive alternative suggestion.

**Reply by Andre601** [reply]
Since the idea is incooperated into a PR with some additional text would it perhaps not be a wrong idea to mention those points you just said, so that people are aware of it.
